packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -614,6 +656,111 @@ function updateWorkInProgressHook(): Hook {
   return workInProgressHook;
 }
 
+// use a Symbol to allow for any value including null and undefined to be memoized
+const EMPTY = Symbol('empty');
+
+function mountContextImpl<C>(
+  context: ReactContext<C>,
+  selector?: ObservedBits | (C => any),
+): any {
+  const hook = mountWorkInProgressHook();
+  if (typeof selector === 'function') {
+    let contextValue = readContext(context);
+    let selection = selector(contextValue);
+    hook.memoizedState = {
+      context,
+      contextValue,
+      selector,
+      selection,
+      stashedContextValue: EMPTY,
+      stashedSelection: EMPTY,
+    };
+    hook.bailout = bailoutContext;
+    return selection;
+  } else {
+    // selector is the legacy observedBits api
+    let observedBits = selector;
+    let contextValue = readContext(context, observedBits);
+    hook.memoizedState = {
+      context,
+      contextValue,
+      selector: null,
+      selection: EMPTY,
+      stashedContextValue: EMPTY,
+      stashedSelection: EMPTY,
+    };
+    hook.bailout = bailoutContext;
+    return contextValue;
+  }
+}
+
+function updateContextImpl<C>(
+  context: ReactContext<C>,
+  selector?: ObservedBits | (C => any),
+): any {
+  const hook = updateWorkInProgressHook();
+  const memoizedState = hook.memoizedState;
+  if (memoizedState.stashedSelection !== EMPTY) {
+    // context and selector could not have changed since we attempted a bailout
+    memoizedState.contextValue = memoizedState.stashedContextValue;
+    memoizedState.selection = memoizedState.stashedSelection;
+    memoizedState.stashedContextValue = EMPTY;
+    memoizedState.stashedSelection = EMPTY;
+    // need to readContext to reset dependency
+    readContext(context);
+    return memoizedState.selection;
+  }
+  memoizedState.context = context;
+  if (typeof selector === 'function') {
+    let selection = memoizedState.selection;
+    let contextValue = readContext(context);
+
+    // when using a selector only recomput if the context value is different or
+    // the selector is different than the memoized state
+    if (
+      contextValue !== memoizedState.contextValue ||
+      selector !== memoizedState.selector
+    ) {
+      selection = selector(contextValue);
+      memoizedState.contextValue = contextValue;
+      memoizedState.selector = selector;
+      memoizedState.selection = selection;
+    }
+    return selection;
+  } else {
+    // selector is actually observedBits
+    let observedBits = selector;
+    let contextValue = readContext(context, observedBits);
+    memoizedState.contextValue = contextValue;
+    return contextValue;
+  }
+}
+
+function bailoutContext(hook: Hook): boolean {
```

Comment:
these bailout functions will be run outside the context of the function component execution itself so they have to be given the hook in consideration